torch/profiler/profiler.py

```diff
@@ -241,16 +397,45 @@ def __init__(
             self.schedule = _default_schedule_fn
             self.record_steps = False
         self.on_trace_ready = on_trace_ready
-        self.record_shapes = record_shapes
-        self.with_flops = with_flops
-        self.profile_memory = profile_memory
-        self.with_stack = with_stack
-        self.with_modules = with_modules
         self.step_num = 0
         self.current_action = self.schedule(self.step_num)
-        self.profiler: Optional[prof.profile] = None
         self.step_rec_fn: Optional[prof.record_function] = None
 
+        self.action_map: Dict[Tuple[ProfilerAction, Optional[ProfilerAction]], List[Any]] = {
+            # key is (prev_action, current_action), value is action list corresponding to the state pair.
+            (ProfilerAction.NONE, ProfilerAction.NONE): [],
+            (ProfilerAction.NONE, ProfilerAction.WARMUP): [self.prepare_trace],
+            (ProfilerAction.NONE, ProfilerAction.RECORD): [self.prepare_trace, self.start_trace],
+            (ProfilerAction.NONE, ProfilerAction.RECORD_AND_SAVE): [self.prepare_trace, self.start_trace],
+            (ProfilerAction.WARMUP, ProfilerAction.NONE): [
+                partial(warn, "Incorrect schedule: WARMUP followed by NONE"),
+                self.start_trace,
+                self.stop_trace],
+            (ProfilerAction.WARMUP, ProfilerAction.WARMUP): [],
+            (ProfilerAction.WARMUP, ProfilerAction.RECORD): [self.start_trace],
+            (ProfilerAction.WARMUP, ProfilerAction.RECORD_AND_SAVE): [self.start_trace],
+            (ProfilerAction.RECORD, ProfilerAction.NONE): [
+                partial(warn, "Incorrect schedule: RECORD followed by NONE"),
+                self.stop_trace],
+            (ProfilerAction.RECORD, ProfilerAction.WARMUP): [
+                partial(warn, "Incorrect schedule: RECORD followed by WARMUP"),
+                self.stop_trace],
+            (ProfilerAction.RECORD, ProfilerAction.RECORD): [],
+            (ProfilerAction.RECORD, ProfilerAction.RECORD_AND_SAVE): [],
+            (ProfilerAction.RECORD_AND_SAVE, ProfilerAction.NONE): [self.stop_trace, self._trace_ready],
+            (ProfilerAction.RECORD_AND_SAVE, ProfilerAction.WARMUP): [self.stop_trace, self._trace_ready, self.prepare_trace],
+            (ProfilerAction.RECORD_AND_SAVE, ProfilerAction.RECORD): [],
```

Comment:
In logic before, the (ProfilerAction.RECORD_AND_SAVE, ProfilerAction.RECORD) should map [self._stop_trace, self._trace_ready, self.prepare_trace, self.start_trace]

Reply:
good catch. yes, your are right. that's is a behavior change  and will make it consistent.